### PR TITLE
Fix broadcasting issues in binary_cross_entropy_with_logits

### PR DESCRIPTION
**Pull request by alykhantejani**
In response to issue #1939 where this failed, due to broadcasting because `o` and `t` have different shapes

```Python
import torch
import torch.nn as nn
from torch.autograd import Variable
import numpy as np

sigmoid = nn.Sigmoid()

t = np.round(np.random.rand(64))
o = np.random.rand(64,1) - 0.5

t = Variable(torch.Tensor(t))
o = Variable(torch.Tensor(o))

print(nn.BCEWithLogitsLoss()(o, t))
print(nn.BCELoss()(sigmoid(o), t)) # Different numbers
```

### CODE REVIEW

**Review by apaszke — COMMENTED** [review]
I don't think it's a good solution. It's better to have a input/target shape spec

**Comment by alykhantejani** [comment]
@apaszke you mean assert that ` target.is_same_size(input)`?

**Comment by apaszke** [comment]
Yeah I guess that if it's an element-wise loss then they should match exactly.

**Comment by alykhantejani** [comment]
This was my first thought too (as the docstring also says they should be the same size - the same for `BCELoss`).

cc @martinarjovsky who had some thoughts on this.

**Comment by alykhantejani** [comment]
(btw I think there's still an issue in my weight sizing in the test - I'll fix that once we reach a conclusion on how we should fix this issue generally)

**Comment by alykhantejani** [comment]
@apaszke if we add a check in `binary_cross_entropy_with_logits` to raise if `target.size() != input.size()`, then we should also add this to `binary_cross_entropy` as their interfaces should match. 

However, this will probably break backwards compatibility as I think many people use it in the following way:
```Python
target = Variable(torch.rand(64))
output = Variable(torch.rand(64,1) - 0.5)
nn.BCELoss()(sigmoid(output), target)
```

**Comment by alykhantejani** [comment]
Looking at the interface of `BCELoss` a bit more, it is also quite inconsistent i.e.
```
target = Variable(torch.rand(5))
output = Variable(torch.rand(5, 1) - 0.5)
nn.BCELoss()(nn.Sigmoid()(output), target) # this works fine
weight = torch.ones(1)
nn.BCELoss(weight)(nn.Sigmoid()(output), target) #this fails as target.size() != input.size()
nn.BCELoss(weight)(nn.Sigmoid()(output), target.unsqueeze(1)) #this works
```

I'm in favour of asserting `target.size() == input.size()` as it leads to simpler code, happy to ammend this PR if you agree @apaszke  (but it will break backwards compatibility of `BCELoss` with single dim targets)

**Comment by apaszke** [comment]
How about adding strict checks to `with_logits` version and deprecation warnings to the ones that we had earlier?

**Comment by alykhantejani** [comment]
Sounds sensible - I'll update this PR with that then

**Comment by alykhantejani** [comment]
@apaszke I've made the changes and added the warning in `functional.binary_cross_entropy` and in `torch/nn/_functions/thnn/loss.BCELoss`. Should I just add the warning in `BCECriterion.c` instead?

**Comment by martinarjovsky** [comment]
Hey! Adding checks for sizes would be good. The thing I think it's really necessary is that behaviour in the 'with_logits' version is exactly the same as the one on probabilities [for BCE and non-binary CE]. The main thing that was worrisome to me was that a shape mismatch affected one and not the other. Cheers :)

**Review by soumith — APPROVED** [review]

**Comment by soumith** [comment]
thanks Alykhan!

File: torch/nn/functional.py [hunk 1]
```diff
@@ -670,6 +670,10 @@ def binary_cross_entropy(input, target, weight=None, size_average=True):
                 sizeAverage is set to False, the losses are instead summed
                 for each minibatch.
     """
+    if not target.is_same_size(input):
+        warnings.warn("Using a target size ({}) that is different to the input size ({}) is deprecated. "
```
Comment (fmassa) [inline torch/nn/functional.py]: Just a question: won't this function raise the warnings twice, one in the functional interface, and the other in the `_functions`?

Comment (alykhantejani) [inline torch/nn/functional.py]: @fmassa I believe the `BCELoss` module calls `BCECriterion.updateOutput` directly and not the function in `nn.functional` see here: https://github.com/pytorch/pytorch/blob/master/torch/nn/_functions/thnn/loss.py#L28  and here: https://github.com/pytorch/pytorch/blob/master/torch/nn/_functions/thnn/auto.py#L266
